plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+import asyncio
+
+from django.core.files import File
+from django.core.files.storage import default_storage
+from django.db.models import Q
+
+from pulpcore.plugin.models import Artifact, ProgressBar
+
+
+async def query_existing_artifacts(in_q, out_q):
+    """
+    Stages API stage that replaces :attr:`DeclarativeContent.content` objects with already-saved
+    :class:`~pulpcore.plugin.models.Artifact` objects.
+
+    This stage expects :class:`~pulpcore.plugin.stages.DeclarativeContent` units from `in_q` and
+    inspects their associated :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects. Each
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object stores one
+    :class:`~pulpcore.plugin.models.Artifact`.
+
+    This stage inspects any unsaved :class:`~pulpcore.plugin.models.Artifact` objects and searches
+    using their metadata for existing saved :class:`~pulpcore.plugin.models.Artifact` objects inside
+    Pulp with the same digest value(s). Any existing :class:`~pulpcore.plugin.models.Artifact`
+    objects found will replace their unsaved counterpart in the
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object.
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after all of its
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects have been handled.
+
+    This stage drains all available items from `in_q` and batches everything into one large call to
+    the db for efficiency.
+
+    Args:
+        in_q (:class:`asyncio.Queue`): The queue to receive
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+        out_q (:class:`asyncio.Queue`): The queue to put
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+
+    Returns:
+        The query_existing_artifacts stage as a coroutine to be included in a pipeline.
+    """
+    declarative_content = []
+    shutdown = False
+    while True:
+        try:
+            content = in_q.get_nowait()
+        except asyncio.QueueEmpty:
+            if not declarative_content:
+                content = await in_q.get()
+                declarative_content.append(content)
+                continue
+        else:
+            declarative_content.append(content)
+            continue
+
+        all_artifacts_q = Q(pk=None)
+        for content in declarative_content:
+            if content is None:
+                shutdown = True
+                continue
+
+            for declarative_artifact in content.d_artifacts:
+                one_artifact_q = Q()
```

Comment:
This should be:
```
Q(pk=None)
```
to ensure you don't match everything in the case where a `DeclarativeArtifact` is queued with none of the digests set.

Reply:
good call. will fix in next push.

Reply:
I can't just use `Q(pk=None)` because this is being combined with an AND query. I just tested and Q() == False and Q() | Q(anything=anything) is True, so I'm going with this patch. Let me know if I should do something different.

```
diff --git a/plugin/pulpcore/plugin/stages/artifact_stages.py b/plugin/pulpcore/plugin/stages/artifact_stages.py
index 64e78609b..aedce7808 100644
--- a/plugin/pulpcore/plugin/stages/artifact_stages.py
+++ b/plugin/pulpcore/plugin/stages/artifact_stages.py
@@ -65,7 +65,8 @@ async def query_existing_artifacts(in_q, out_q):
                     if digest_value:
                         key = {digest_name: digest_value}
                         one_artifact_q &= Q(**key)
-                all_artifacts_q |= one_artifact_q
+                if one_artifact_q:
+                    all_artifacts_q |= one_artifact_q
 
         for artifact in Artifact.objects.filter(all_artifacts_q):
             for content in batch:
```